sycl/include/CL/sycl/intel/sub_group.hpp

```diff
@@ -25,69 +26,167 @@ template <typename T, access::address_space Space> class multi_ptr;
+                  Space == access::address_space::global_space>;
+
+template <typename T, access::address_space Space>
+T load(const multi_ptr<T, Space> src) {
+  using Ty = SelectBlockTy<T>;
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  Ty Ret = __spirv_SubgroupBlockReadINTEL<Ty>(
+      reinterpret_cast<GlobalPtrTy>(src.get()));
+
+  // TODO: any better representation of reinterpret cast from Ty to T for Ret?
+  return *((T *)&Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_integral<T>::value, T>::type
-calc(T x, intel::plus<T> op) {
-  return __spirv_GroupIAdd<T>(__spv::Scope::Subgroup, O, x);
+template <int N, typename T, access::address_space Space>
+vec<T, N> load(const multi_ptr<T, Space> src) {
+  using Ty = SelectBlockTy<T>;
+  using VecTy = Ty __attribute__((ext_vector_type(N)));
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  VecTy Ret = __spirv_SubgroupBlockReadINTEL<VecTy>(
+      reinterpret_cast<GlobalPtrTy>(src.get()));
+
+  // TODO: any better representation of reinterpret cast from VecTy to vec<T,
+  // N>::vector_t?
+  return *((typename vec<T, N>::vector_t *)&Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<detail::is_floating_point<T>::value, T>::type
-calc(T x, intel::plus<T> op) {
-  return __spirv_GroupFAdd<T>(__spv::Scope::Subgroup, O, x);
+template <typename T, access::address_space Space>
+void store(multi_ptr<T, Space> dst, const T &x) {
+  using Ty = SelectBlockTy<T>;
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  __spirv_SubgroupBlockWriteINTEL(reinterpret_cast<GlobalPtrTy>(dst.get()),
+  // TODO: any better representation of reinterpret cast from T to Ty for x?
+                                      *((Ty *)&x));
 }
 
+template <int N, typename T, access::address_space Space>
+void store(multi_ptr<T, Space> dst, const vec<T, N> &x) {
+  using Ty = SelectBlockTy<T>;
+  using VecTy = Ty __attribute__((ext_vector_type(N)));
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  __spirv_SubgroupBlockWriteINTEL(reinterpret_cast<GlobalPtrTy>(dst.get()),
+  // TODO: any better representation of reinterpret cast from T to Ty for x?
+                                      *((VecTy *)&x));
+}
+
+struct GroupOpISigned {}; struct GroupOpIUnsigned {}; struct GroupOpFP {};
+
+template <class T, typename = void> struct GroupOpTag;
+
+template <class T>
+struct GroupOpTag<T, detail::enable_if_t<detail::is_sigeninteger<T>::value>> {
+  using type = GroupOpISigned;
+};
+
+template <class T>
+struct GroupOpTag<T, detail::enable_if_t<detail::is_sugeninteger<T>::value>> {
+  using type = GroupOpIUnsigned;
+};
+
+template <class T>
+struct GroupOpTag<T, detail::enable_if_t<detail::is_sgenfloat<T>::value>> {
+  using type = GroupOpFP;
+};
+
+#define CALC_BODY(Operation)                                                   \
```

Comment:
I don't think using a macro is necessary here. The function can be written once rather than many times. All that is needed is to create a compile time lookup table from GroupOp and Op to the correct `__spirv_Group...`. Something along the lines of:

```c++
tempate<class, class> struct SpirvGroupFn;
template<T> SpirvGroupFn<GroupOpISigned, intel::minimum<T>> { 
   static constexpr auto value = __spirv_GroupSMin;
};
```

Reply:
`value` is a function pointer here and therefore, this method cannot be used here

Reply:
Yes it is a function pointer but what's the problem? Because it is a `constexpr` function pointer there is no difference to a direct call even at `O0`. With C++20 it is possible to use a type based solution: https://godbolt.org/z/A3vraY. But that's actually worse at O0. Also we can't rely on C++20 and the feature (lambdas in unevaluated context) isn't supported in clang yet.

Reply:
> Yes it is a function pointer but what's the problem? Because it is a constexpr function pointer there is no difference to a direct call even at O0

At LLVM IR level, probably yes, but the check for indirect calls in `SemaSYCL.cpp` is not able to detect this right now.

Reply:
Good point. And the check is what's required by the spec. I'll suggest a change to the spec that spec that constant expression function pointers (/references) are excluded from the rule for indirect calls.